vulpes/upgrades/gamestate.cpp

```diff
@@ -0,0 +1,423 @@
+    {"teams", TABLE_TEAMS_UPGR_MAX, true},
+    {"contrail", TABLE_CONTRAIL_UPGR_MAX, true},
+    {"contrail point", TABLE_CONTRAIL_POINT_UPGR_MAX, true},
+    {"particle", TABLE_PARTICLE_UPGR_MAX, true},
+    {"effect", TABLE_EFFECT_UPGR_MAX, true},
+    {"effect location", TABLE_EFFECT_LOCATION_UPGR_MAX, true},
+    {"particle systems", TABLE_PARTICLE_SYSTEMS_UPGR_MAX, true},
+    {"particle system particles", TABLE_PARTICLE_SYSTEM_PARTICLES_UPGR_MAX, true},
+    {"object looping sounds", TABLE_OBJECT_LOOPING_SOUNDS_UPGR_MAX, true},
+    {"actor", TABLE_ACTOR_UPGR_MAX, true},
+    {"swarm", TABLE_SWARM_UPGR_MAX, true},
+    {"swarm component", TABLE_SWARM_COMPONENT_UPGR_MAX, true},
+    {"prop", TABLE_PROP_UPGR_MAX, true},
+    {"encounter", TABLE_ENCOUNTER_UPGR_MAX, true},
+    {"ai pursuit", TABLE_AI_PERSUIT_UPGR_MAX, true},
+    {"object list header", TABLE_OBJECT_LIST_HEADER_UPGR_MAX, false},
+    {"list object reference", TABLE_LIST_OBJECT_REFERENCE_UPGR_MAX, false},
+    {"hs thread", TABLE_HS_THREAD_UPGR_MAX, true},
+    {"hs globals", TABLE_HS_GLOBALS_UPGR_MAX, true},
+    {"recorded animations", TABLE_RECORDED_ANIMATIONS_UPGR_MAX, true},
+    {"", 0, false}, // Terminate
+};
+
+// Wonky argument order needed for the assembly wrapper to stay simple.
+extern "C" __attribute__((regparm(1)))
+GenericTable* gamestate_table_new_replacement(uint32_t element_size,
+        const char* name, uint16_t element_max) {
+
+    bool found = false;
+    bool use_upgrade_memory = false;
+    size_t i = 0;
+    while (!found && TABLE_UPGRADES[i].new_max != 0) {
+        if (strncmp(name, TABLE_UPGRADES[i].name, 31) == 0) {
+            // Set max to new max.
+            if (TABLE_UPGRADES[i].new_max > element_max) {
+                // Only use our limit if it is higher. Other mods may increase
+                // limits more than us and we have extra memory allocated for
+                // the occasion.
+                element_max = TABLE_UPGRADES[i].new_max;
+            }
+            use_upgrade_memory = TABLE_UPGRADES[i].in_upgrade_memory;
+            found = true;
+            break;
+        }
+        i++;
+    }
+
+    uintptr_t* mem_start;
+    uintptr_t* mem_used;
+    const uintptr_t* mem_max;
+
+    if (use_upgrade_memory) {
+        mem_start = reinterpret_cast<uintptr_t*>(&gamestate_extension_buffer);
+        mem_used  = &used_extension_memory;
+        mem_max   = &ALLOCATED_UPGRADE_MEMORY;
+    } else {
+        mem_start = game_state_globals;
+        mem_used  = game_state_globals_cpu_allocation_size;
+        mem_max   = game_state_globals_buffer_size;
+    }
+
+    // Get the next available spot in vanilla gamestate.
+    auto new_table = reinterpret_cast<GenericTable*>(*mem_start + *mem_used);
+
+    // Calculate and store the total used.
+    *mem_used = *mem_used + sizeof(GenericTable) + element_size * element_max;
+
+    // Make sure the limits are correctly enforced.
+    assert(*mem_used <= *mem_max);
```

Comment:
In a windows environment it will pop up a dialog saying exactly what line it failed on. And has a debugging  function. I eventually want to make the debug stuff more robust before we release this into the wild. I have some plans that will make error reporting a lot more useful.